compiler/rustc_mir_build/src/thir/pattern/const_to_pat.rs

```diff
@@ -272,12 +272,14 @@ impl<'a, 'tcx> ConstToPat<'a, 'tcx> {
 
         let kind = match cv.ty.kind() {
             ty::Float(_) => {
-                tcx.struct_span_lint_hir(
-                    lint::builtin::ILLEGAL_FLOATING_POINT_LITERAL_PATTERN,
-                    id,
-                    span,
-                    |lint| lint.build("floating-point types cannot be used in patterns").emit(),
-                );
+                if self.include_lint_checks {
```

Comment:
This isn't really new to this PR, but it would be nice if `include_lint_checks` had some documentation on when it's enabled and when not. Currently it seems to be enabled in `rustc_mir_build::thir::pattern::MatchVisitor::lower_pattern` but not in `rustc_mir_build::thir::pattern::pat_from_hir`. `lower_pattern` happens when checking the validity of matches and `pat_from_hir` happens when actually constructing mir. It might be nice to mention at the field definition site that `include_lint_checks` will be true when checking validity and false when building mir.

Reply:
It seems somewhat wasteful that the `ConstToPat` lowering even has to run twice (if it didn't, we wouldn't need the `include_lint_checks` field). However, I don't know the broader design considerations behind all of this, I just developed a targeted fix for the issue you reported, so I'm not the right person to ask for additional documentation here...